backend/app/datasets/service.py

```diff
@@ -90,7 +111,58 @@ def get_dataset(self, id: UUID, user: UserModel) -> DatasetGet:
 
         return dataset
 
+    def search_datasets(
+        self, query: str, limit: int, user: UserModel
+    ) -> Sequence[DatasetsSearch]:
+        load_options = get_dataset_load_options()
+        stmt = (
+            select(
+                DatasetModel,
+                func.ts_rank_cd(
+                    DatasetModel.search_vector,
+                    func.websearch_to_tsquery("english", query),
+                    32,  # 32 normalizes all results between 0-1
+                ).label("rank"),
+            )
+            .options(*load_options)
+            .where(
+                DatasetModel.search_vector.op("@@")(
+                    func.websearch_to_tsquery("english", query)
+                )
+            )
+            .order_by(desc("rank"))
+        )
+        results = self.db.execute(stmt).unique().all()
+        filtered_datasets: list[DatasetsSearch] = []
+        for row in results:
+            if len(filtered_datasets) >= limit:
+                return filtered_datasets
+            dataset = row[0]
+            rank = row[1]
+            if self.is_visible_to_user(dataset, user):
+                dataset.rank = rank
+                filtered_datasets.append(dataset)
+        return filtered_datasets
+
+    def recalculate_search_vector_for(self, dataset_id: UUID) -> int:
+        if settings.SEARCH_INDEXING_DISABLED:
+            logger.debug("Search indexing is disabled, skipping recalculation")
+            return 0
+        sql = recalculate_search_vector_dataset_statement(dataset_id)
+        result = self.db.execute(sql, {"dataset_id": dataset_id})
+        return result.rowcount
+
+    def recalculate_search_vector_datasets(self) -> int:
+        if settings.SEARCH_INDEXING_DISABLED:
+            logger.debug("Search indexing is disabled, skipping recalculation")
+            return 0
+        sql = recalculate_search_vector_datasets_statement()
+        result = self.db.execute(sql)
+        self.db.commit()
```

Comment:
this function is only called explicitly in the seeding script so I allow the commit here